pkg/cmd/status/status.go

```diff
@@ -0,0 +1,616 @@
+			HTMLURL string `json:"html_url"`
+		}
+	}
+}
+
+type SearchResult struct {
+	Type       string `json:"__typename"`
+	UpdatedAt  time.Time
+	Title      string
+	Number     int
+	Repository struct {
+		NameWithOwner string
+	}
+}
+
+type Results []SearchResult
+
+func (rs Results) Len() int {
+	return len(rs)
+}
+
+func (rs Results) Less(i, j int) bool {
+	return rs[i].UpdatedAt.After(rs[j].UpdatedAt)
+}
+
+func (rs Results) Swap(i, j int) {
+	rs[i], rs[j] = rs[j], rs[i]
+}
+
+type StatusGetter struct {
+	Client         *http.Client
+	cachedClient   func(*http.Client, time.Duration) *http.Client
+	Org            string
+	Exclude        string
+	AssignedPRs    []StatusItem
+	AssignedIssues []StatusItem
+	Mentions       []StatusItem
+	ReviewRequests []StatusItem
+	RepoActivity   []StatusItem
+}
+
+func NewStatusGetter(client *http.Client, opts *StatusOptions) *StatusGetter {
+	return &StatusGetter{
+		Client:       client,
+		Org:          opts.Org,
+		Exclude:      opts.Exclude,
+		cachedClient: opts.CachedClient,
+	}
+}
+
+func (s *StatusGetter) CachedClient(ttl time.Duration) *http.Client {
+	return s.cachedClient(s.Client, ttl)
+}
+
+func (s *StatusGetter) ShouldExclude(repo string) bool {
+	return strings.Contains(s.Exclude, repo)
+}
+
+func (s *StatusGetter) CurrentUsername() (string, error) {
+	cachedClient := s.CachedClient(time.Hour * 48)
+	cachingAPIClient := api.NewClientFromHTTP(cachedClient)
+	currentUsername, err := api.CurrentLoginName(cachingAPIClient, ghinstance.Default())
+	if err != nil {
+		return "", fmt.Errorf("failed to get current username: %w", err)
+	}
+
+	return currentUsername, nil
+}
+
+func (s *StatusGetter) ActualMention(n Notification) (string, error) {
+	currentUsername, err := s.CurrentUsername()
+	if err != nil {
+		return "", err
+	}
+
+	// long cache period since once a comment is looked up, it never needs to be
+	// consulted again.
+	cachedClient := s.CachedClient(time.Hour * 24 * 30)
+	c := api.NewClientFromHTTP(cachedClient)
+	resp := struct {
+		Body string
+	}{}
+	if err := c.REST(ghinstance.Default(), "GET", n.Subject.LatestCommentURL, nil, &resp); err != nil {
+		return "", err
+	}
+
+	var ret string
+
+	if strings.Contains(resp.Body, "@"+currentUsername) {
+		ret = resp.Body
+	}
+
+	return ret, nil
+}
+
+// These are split up by endpoint since it is along that boundary we parallelize
+// work
+
+// Populate .Mentions
+func (s *StatusGetter) LoadNotifications() error {
+	perPage := 100
+	c := api.NewClientFromHTTP(s.Client)
+	query := url.Values{}
+	query.Add("per_page", fmt.Sprintf("%d", perPage))
+	query.Add("participating", "true")
+	query.Add("all", "true")
+
+	// this sucks, having to fetch so much :/ but it was the only way in my
+	// testing to really get enough mentions. I would love to be able to just
+	// filter for mentions but it does not seem like the notifications API can
+	// do that. I'd switch to the GraphQL version, but to my knowledge that does
+	// not work with PATs right now.
```

Comment:
To avoid having to suffer from N+1 query problem with REST, one option could be to use the GraphQL API that's currently available only to our OAuth app and otherwise print some non-fatal notice about this being unavailable to other types of tokens, including PATs. Could that be an acceptable compromise?

Reply:
We had chatted about this and I decided I wasn't happy with the PAT limitation; when this API is more widely available, that would be a great improvement though.